### ISSUE TITLE
ob1 pml with smcuda btl doesn't work with slurm gpu binding

### ISSUE THREAD

**Issue by simonbyrne** [open-mpi/ompi#11949]
### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)
4.1.5

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

Built from source with UCX and CUDA, PMI and Slurm support.

### Please describe the system on which you are running

* Operating system/version: Ubuntu
* Computer hardware: Multi GPU (NVIDIA A100-SXM4)
* Network type: single node

-----------------------------

## Details of the problem

If I launch a job with `srun -n 2 --gpus-per-task=1` (which uses Slurm GPU binding), then I get the following warning:

```
--------------------------------------------------------------------------
The call to cuIpcOpenMemHandle failed. This is an unrecoverable error
and will cause the program to abort.
  Hostname:                         clima
  cuIpcOpenMemHandle return value:  1
  address:                          0x7fd865200000
Check the cuda.h file for what the return value means. A possible cause
for this is not enough free device memory.  Try to reduce the device
memory footprint of your application.
--------------------------------------------------------------------------
[clima.gps.caltech.edu:1868819] Failed to register remote memory, rc=-1
```

I don't see it if I switch to using `srun -n 2 --gpus=2` (which doesn't use GPU binding), or Open MPI's launcher.

I also don't see it if I switch to using using the UCX pml (`export OMPI_MCA_pml=ucx`).

A similar problem was also seen here: https://stackoverflow.com/questions/67091056/gpu-allocation-in-slurm-gres-vs-gpus-per-task-and-mpirun-vs-srun

**Comment by rhc54** [comment]
Just posted this on the stackoverflow issue you cited - then realized it had been filed over a year ago:
```
Well, FWIW - variant B1 won't work because mpirun
uses srun under the covers only to launch its daemons.
There is only one daemon/node, and thus srun is only
assigning one GPU to that task (the daemon). The
daemon then fork/exec's the application procs, which
inherit that GPU assignment envar.

Variant A1 only works because you asked for two
GPUs/task, and you happen to run two application
procs on each node. If you had run a 3rd proc, or had
asked for only one GPU/task, it would have failed - for
the same reason as was given above.
```
I cannot speak to the specific problem you cite - just posting the above because it might help you to understand why `mpirun` behaves differently than `srun`.

**Comment by simonbyrne** [comment]
I should clarify that the job still runs, it just outputs a lot of warnings.

**Comment by simonbyrne** [comment]
Okay, following the suggestion in the docs, doing
```
export OMPI_MCA_btl_smcuda_use_cuda_ipc=0
```
also suppresses the warnings.

Also, if I use UCX, and enable UCX debug logging (`UCX_LOG_LEVEL=debug`), then I get
```
[1695770285.441684] [clima:1875346:0]           mpool.c:282  UCX  DEBUG   mpool ucp_rkeys: allocated chunk 0x57f6790 of 16472 bytes with 128 elements
[1695770285.441759] [clima:1875346:0]  cuda_ipc_cache.c:111  UCX  DEBUG   cuIpcOpenMemHandle() failed: invalid argument
[1695770285.441762] [clima:1875346:0]  cuda_ipc_cache.c:307  UCX  DEBUG   dest:1875345:0: failed to open ipc mem handle. addr:0x7fcd05200000 len:8000
```

Which suggests that CUDA IPC can't generally be used with Slurm GPU binding. What transport mechanism is actually used in that case?

**Comment by simonbyrne** [comment]
> What transport mechanism is actually used in that case?

To answer my own question, I profiled it and it looks like in both cases it copies via this host (I see `CUDA memcpy DtoH` and `CUDA memcpy HtoD` operations).

**Comment by BenWibking** [comment]
CUDA IPC can only be used when all devices on a node are visible (in the sense of `CUDA_VISIBLE_DEVICES`) to all of the MPI ranks on a node. If not, it can only do copies buffered by host memory, which severely reduces performance. Unfortunately, this is not documented anywhere: https://github.com/open-mpi/ompi/issues/10871

**Comment by simonbyrne** [comment]
That's what I had thought too, but `CUDA_VISIBLE_DEVICES` always reports `0` when using GPU binding:
```
├ srun -n 2 --gpus-per-task=1 printenv CUDA_VISIBLE_DEVICES
0
0
```
however these are actually different devices
```
├ srun -n 2 --gpus-per-task=1 nvidia-smi -L
GPU 0: NVIDIA A100-SXM4-80GB (UUID: GPU-09c6a428-f4ea-8730-fd1c-f3d323d0e161)
GPU 0: NVIDIA A100-SXM4-80GB (UUID: GPU-02743130-9954-abbd-7a10-877b693748d6)
```

It turns out that our Slurm setup is configured to use cgroups to control device allocations, i.e. `/etc/slurm/cgroup.conf` has the line
```
ConstrainDevices=yes
```

So I think the summary is that if users want GPU direct communication, they should not use Slurm GPU binding.

I agree it would be better if this were documented somewhere (and if the warning was clearer that it was falling back on host-buffered copies)

**Comment by BenWibking** [comment]
Yes, that's exactly what I have found on Slurm clusters that I've used. I have to set `--gpu-bind=none` in order to get CUDA IPC to be used as the intra-node transport.

**Comment by jyvet** [comment]
Right, this issue is likely observed with all MPI implementations (I confirm, same issue with MPICH) and all GPU types. The best which can be done is documenting the outcome (this is already the case with Cray MPI).

Keep in mind that if you use `--gpu-bind=none` and manually set `CUDA_VISIBLE_DEVICES` (or `ROCR_VISIBLE_DEVICES`), core affinites should ideally be preserved to avoid NUMA or NUIOA effects. In other words, Slurm knows which GPU to assign to which subset of CPU cores for a given task. You'll have to get that mapping right manually. 

I'm wondering whether a workaround should be implemented in Slurm. Might be a naive question, but is it really useful to set different cgroup constraints on GPUs between tasks of the same job? Slurm could use cgroups to isolate all resources allocated by the job, then expose proper GPU(s) (relying on affinities defined in gres.conf) per task using  CUDA_VISIBLE_DEVICES (or ROCR_VISIBLE_DEVICES) environment variable.

**Comment by BenWibking** [comment]
I have only seen this issue with NVIDIA GPUs. I don't see this issue when using SLURM bindings `--gpu-bind=closest` with OpenMPI with AMD GPUs.

**Comment by BenWibking** [comment]
> Are Slurm cgroups enabled with the proper Slurm GRES configuration for AMD GPUs (some HPC sites don't enable it)? Is IPC enabled?
> 
> I can tell Cray MPI with AMD MI250X GPUs is impacted. If OpenMPI is impacted with NVIDIA GPUs, I don't see any reason why it wouldn't be the case with AMD GPUs.

On Frontier, `--gpu-bind=closest` still allows IPC between MI250X GPUs with Cray MPI (I have checked). I don't know what the GRES configuration is on Frontier.

On a non-Cray cluster with MI100 GPUs, `--gpu-bind=closest` still allows IPC in my tests.

ROCm and CUDA are completely different software stacks, so I don't see why there couldn't be a difference.

**Comment by simonbyrne** [comment]
> Might be a naive question, but is it really useful to set different cgroup constraints on GPUs between tasks of the same job? Slurm could use cgroups to isolate all resources allocated by the job, then expose proper GPU(s) (relying on affinities defined in gres.conf) per task using  `CUDA_VISIBLE_DEVICES` (or `ROCR_VISIBLE_DEVICES`) environment variable.

I agree that something like this would make more sense. 

I have opened a ticket here: https://bugs.schedmd.com/show_bug.cgi?id=17793, unfortunately my institute doesn't have a Slurm support contract, so I suspect not much will happen. If someone with a support contract wants it fixed, please let them know.

**Comment by jyvet** [comment]
> On Frontier, `--gpu-bind=closest` still allows IPC between MI250X GPUs with Cray MPI (I have checked). I don't know what the GRES configuration is on Frontier.
> 
> On a non-Cray cluster using OpenMPI with MI100 GPUs, `--gpu-bind=closest` still allows IPC in my tests.
> 
> ROCm and CUDA are completely different software stacks, so I don't see why there couldn't be a difference.

Ah this is Frontier. I don't think there is any cgroup enforcement there. This is why you cannot reproduce.

You can type `man mpi` with Cray MPI, this is documented. The issue is related to cgroups. GPUs isolated in different cgroups cannot initiate P2P communications.

**Comment by jyvet** [comment]
FYI a SchedMD case opened with a proper support contract: 
https://bugs.schedmd.com/show_bug.cgi?id=17875

No luck so far... Maybe other sites with the same issue should upvote that case.
